docs/gke/iap.md

```diff
@@ -0,0 +1,272 @@
+when using GKE.
+
+  * IAP allows you to control access to JupyterHub using Google logins
+  * Using IAP secures access to JupyterHub using HTTPs
+  * IAP allows users to safely and easily connect to JupyterHub
+  * You will need your own domain which can be purchased and managed using Google domains or another provider
+
+If you aren't familiar with IAP you might want to start by looking at those docs.
+
+### Preliminaries
+
+##### Create an external static IP address
+
+```
+gcloud compute --project=${PROJECT} addresses create kubeflow --global
+```
+
+Use your DNS provider (e.g. Google Domains) create a type A custom resource record that associates the host you want e.g "kubeflow"
+with the IP address that you just reserved.
+  * Instructions for [Google Domains](https://support.google.com/domains/answer/3290350?hl=en&_ga=2.237821440.1874220825.1516857441-1976053267.1499435562&_gac=1.82147044.1516857441.Cj0KCQiA-qDTBRD-ARIsAJ_10yKS7G1HPa1aoM8Mk_4VagV9wIi5uKkMp5UWJGDNejKxWPKUO_A6ri4aAsahEALw_wcB)
+
+##### Create a self signed certificate
+
+The certificate is needed for HTTPs
+  
+```
+ENDPOINT_URL=${HOST}.${YOURDOMAIN}
+mkdir -p ~/tmp/${ENDPOINT_URL}
+TLS_KEY_FILE=~/tmp/${ENDPOINT_URL}/tls.key
+TLS_CRT_FILE=~/tmp/${ENDPOINT_URL}/tls.crt
+
+openssl req -x509 -nodes -days 365 -newkey rsa:2048 \
+  -subj "/CN=${ENDPOINT_URL}/O=Google LTD./C=US" \
+  -keyout ${TLS_KEY_FILE} -out ${TLS_CRT_FILE}
+```
+  
+  * HOST will be the value of the custom resource record that you created to map to your IP address.
+
+Create a K8s secret to store the SSL certificate.
+
+```
+SECRET_NAME=<Name for your secret.>
+kubectl -n ${NAMESPACE} create secret generic ${SECRET_NAME}  --from-file=${TLS_KEY_FILE} --from-file=${TLS_CRT_FILE}
+```
+
+
+#### Create oauth client credentials
+
+Create an OAuth Client ID to be used to identify IAP when requesting acces to user's email to verify their identity.
+
+1. Set up your OAuth consent screen:
+* Configure the [consent screen](https://console.cloud.google.com/apis/credentials/consent).
+* Under Email address, select the address that you want to display as a public contact. You must use either your email address or a Google Group that you own.
+* In the Product name box, enter a suitable like save `kubeflow`
+* Click Save.
+1.  Click Create credentials, and then click OAuth client ID.
+  * Under Application type, select Web application. In the Name box enter a name, and in the Authorized redirect URIs box, enter 
+
+```
+  https://${ENDPOINT_URL}/_gcp_gatekeeper/authenticate, 
+```
+
+1. After you enter the details, click Create. Make note of the **client ID** and **client secret** that appear in the OAuth client window because we will
+   need them later to enable IAP.
+
+Save the OAuth client ID and secret to variables for later use:
+
+### Setup Ingress
+
+```
+ks generate ${ENVIRONMENT} iap-ingress --secretName=${SECRET_NAME} --ipName=${IP_NAME} --namespace=${NAMESPACE}
```

Comment:
ksonnet [concepts doc](https://github.com/ksonnet/ksonnet/blob/master/docs/concepts.md#environment) says that environments are defined under environments/ which doesn't exist yet?

Reply:
I added instructions explaining that you should have initialized your ksonnet app already by following the user guide. So the environments directory should exist.